### PR TITLE
Make Territory.getUnits() return an unmodifiable list. (Reland)

### PR DESCRIPTION
**Pull request by asvitkine**
Make Territory.getUnits() return an unmodifiable list. (Reland)

Previously, it was returning a copy, which was actually quite expensive given how often this was called. On a Domination 1914 game, with all AIs, this took 100 seconds of CPU time (across different threads since it's used in battle sim) over a single round.

A few places that need a copy are updated to do the copy on their side.

## Functional Changes

[] New map or map update
[] New Feature
[] Feature update or enhancement
[] Feature Removal
[] Code Cleanup or refactor
[] Configuration Change
[] Problem fix: 
[X] Other:   Optimization

## Testing
Ran ./verify

### CODE REVIEW

**Comment by DanVanAtta** [comment]
@asvitkine could you try creating this PR on a different branch. travis-ci.org is not picking this up. I"m not sure what's up with that service, it seems there are issues after the migration to  travis-ci.com

File: game-core/src/main/java/games/strategy/triplea/delegate/AbstractPlaceDelegate.java [hunk 1]
```diff
@@ -1652,14 +1652,13 @@ private boolean getCanAllUnitsWithRequiresUnitsBePlacedCorrectly(
...
     }
-    final Collection<Unit> unitsInTo = to.getUnits();
     final Collection<Unit> unitsPlacedAlready = getAlreadyProduced(to);
     if (Matches.territoryIsWater().test(to)) {
       for (final Territory current : getAllProducers(to, player, null, true)) {
         unitsPlacedAlready.addAll(getAlreadyProduced(current));
       }
     }
-    final Collection<Unit> unitsAtStartOfTurnInTo = new ArrayList<>(unitsInTo);
+    final Collection<Unit> unitsAtStartOfTurnInTo = new ArrayList<>(to.getUnits());
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/AbstractPlaceDelegate.java:1661]: Would it be simpler to have the copy be done once and consistently in the source method?


EG: UnitCollection.java
```
  default Collection<Unit> getUnits() {
    return getUnitCollection().getUnits();
  }
```

**Comment by DanVanAtta** [comment]
@asvitkine a second thing to try, it looks like your branches are consistently not being built, try to enable travis for your forked repository.

We need PRs submitted by you to appear in one of these two:
- https://travis-ci.org/github/triplea-game/triplea/pull_requests
- https://travis-ci.com/github/triplea-game/triplea/pull_requests

I'm wondering if there needs to be permissions into your fork to enable this build.

File: game-core/src/main/java/games/strategy/triplea/delegate/AbstractPlaceDelegate.java (hunk 1 above)
Comment (asvitkine) [reply game-core/src/main/java/games/strategy/triplea/delegate/AbstractPlaceDelegate.java:1661]: The whole point is to not copy.

The places that are changing are the few places that still need to copy, but those are the ones we don't care about. The point is all the other places are now *not* copying - and those are the ones that were contributing to the run time.

Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/triplea/delegate/AbstractPlaceDelegate.java:1661]: Sorry, mistyped, meant to suggest we do the unmodifiable list copy here.

It's not super surprising we see some issues when copying, the usages of this method all seem to assume a constant time operation and invoke the method all over the place, even if the list is not expected to change.

**Comment by DanVanAtta** [comment]
@asvitkine please see if adding travis permissions to your fork fixes the PR issue. I suspect strongly the migration to travis-ci.com caused that problem. If the permissions fixes the problem then we'll have a short term solution and more information to supply to them.

**Comment by DanVanAtta** [comment]
@asvitkine this URL should hopefully give you the right prompts: https://travis-ci.com/github/asvitkine/triplea/

**Comment by DanVanAtta** [comment]
Exception after the build:

```
games.strategy.triplea.delegate.MoveDelegateTest > testAirDifferingRouts() FAILED

    java.lang.UnsupportedOperationException

        at java.base/java.util.Collections$UnmodifiableCollection.removeAll(Collections.java:1071)

        at games.strategy.triplea.delegate.AirMovementValidator.validateAirCaughtByMovingCarriersAndOwnedAndAlliedAir(AirMovementValidator.java:349)

        at games.strategy.triplea.delegate.AirMovementValidator.validateAirCanLand(AirMovementValidator.java:156)

        at games.strategy.triplea.delegate.MoveValidator.validateMove(MoveValidator.java:104)

        at games.strategy.triplea.delegate.MoveDelegate.performMove(MoveDelegate.java:623)

        at games.strategy.triplea.delegate.AbstractMoveDelegate.move(AbstractMoveDelegate.java:117)

        at games.strategy.triplea.delegate.MoveDelegateTest.testAirDifferingRouts(MoveDelegateTest.java:1207)
```

File: game-core/src/main/java/games/strategy/triplea/delegate/AbstractPlaceDelegate.java (hunk 1 above)
Comment (asvitkine) [reply game-core/src/main/java/games/strategy/triplea/delegate/AbstractPlaceDelegate.java:1661]: I'm still not following your suggestions.

"meant to suggest we do the unmodifiable list copy here"

The next line modifies the list copy, so here it needs be modifiable.

**Comment by asvitkine** [comment]
> https://travis-ci.com/github/asvitkine/triplea/

Gives me an error:

"We couldn't display the repository
asvitkine/triplea
This repository may not exist or you may need to use another account to have access to it"

**Comment by asvitkine** [comment]
However, when I log in to travis, I see "The account asvitkine has been flagged for potential abuse. Please contact support@travis-ci.com."

Since it used to work before, I wonder if they saw too much activity from me or something and blocked it? I'll reach out to them, but that's kind of silly.

**Pull request merged**